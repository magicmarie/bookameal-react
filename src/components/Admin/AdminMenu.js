import React, { Component } from "react";
import Loader from "react-loader";
import { notify } from "react-notify-toast";
import axiosInstance from "../common/Apicalls";
import Menu from "./Menu";
import WeekDays from "../common/WeekDays";
import getCurrentDay from "../common/CurrentDay";

/**
 *
 *@param {string} day
 * @class AdminMenu
 * @extends {Component}
 */
class AdminMenu extends Component {
  state = {
    menus: [],
    loaded: false,
    currentMenu: {
      meals: [],
      name: ""
    }
  };

  /**
   *@returns {menus} menus
   * @memberof AdminMenu
   */
  componentDidMount() {
    this.getMenus();
  }

  // get menu: admin side
  /**
   *@returns {null} null
   * @memberof AdminMenu
   */
  getMenus = () => {
    axiosInstance
      .get("/admin-menus")
      .then(response => {
        const today = getCurrentDay();
        const currentMenu = response.data.menus.find(
          menu => today === menu.name
        );
        this.setState({
          menus: response.data.menus,
          loaded: true,
          currentMenu
        });
      })
      .catch(error => {
        if (error.response) {
          const { status } = error.response;
          if (status === 401) {
            localStorage.removeItem("token");
          }
        } else if (error.request) {
          notify.show("Wrong request", "warning", 2500);
        }
      });
  };

  getMenu = day => {
    day = day.charAt(0).toUpperCase() + day.slice(1);
    // manipulate the menus array from the state to get currentMenu
    const { menus } = this.state;
    const currentMenu = menus.find(menu => day === menu.name);
    this.setState({ currentDay: day, currentMenu });
  };
  // delete meal from the menu: close modal, show success message, get the menu
  ConfirmDeleteMeal = id => {
    axiosInstance.delete(`/menu/${id}`).then(response => {
      notify.show(response.data.message, "success", 2500);
      document.getElementById(`cancelModal${id}`).click();
      this.getMenus();
    });
  };

  /**
   * @returns {any} rendered items
   * @memberof AdminMenu
   */
  render() {
    const { currentMenu, loaded } = this.state;
    const menuDetails =
      currentMenu.meals.length === 0 ? (
        <div>No Menu Found</div>
      ) : (
        currentMenu.meals.map(menu => (
          <div className="col-md-3 col-lg-3" key={menu.id}>
            <Menu
              id={menu.id}
              mealName={menu.meal_name}
              price={menu.price}
              key={menu.id}
              ConfirmDeleteMeal={this.ConfirmDeleteMeal}
            />
          </div>
        ))
      );
    return (
      <div className="row">
        <div className="col-md-2">
          <div className="header">Days</div>
          <ul className="list-group">
            <WeekDays getMenu={this.getMenu} />
          </ul>
        </div>
        <div className="vl" />
        <div className="col-md-9 meal-list">
          <h2 className="header text-center">
            {currentMenu.name}
            's Menu List
          </h2>
          <Loader loaded={loaded}>
            <div className="row">{menuDetails}</div>
          </Loader>
        </div>
      </div>
    );
  }
}
export default AdminMenu;
