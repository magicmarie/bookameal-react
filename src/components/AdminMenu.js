import React, { Component } from "react";
import Loader from "react-loader";
import axiosInstance from "./common/Apicalls";
import { notify } from "react-notify-toast";
import Menu from "./Menu";

class AdminMenu extends Component {
  state = {
    menu: [],
    loaded: false
  };
  // get the menu
  getMenu = () => {
    axiosInstance
      .get("/menu")
      .then(response => {
        const menu = response.data.Menu;
        this.setState({ menu, loaded: true });
      })
      .catch(error => {
        if (error.response) {
          const { status } = error.response;
          if (status === 404) {
            this.setState({
              menu: []
            });
          } else if (status === 401) {
            localStorage.removeItem("token");
          }
        } else if (error.request) {
          notify.show("Wrong request", "warning", 4000);
        }
      });
  };
  // make an order from the menu
  handleAddOrder = (id, meal_id, meal_name, price) => {
    axiosInstance
      .post(`/orders/${id}/${meal_id}`)
      .then(response => {
        const orders = response.data;
        this.setState({ orders });
        notify.show(response.data.message, "success", 4000);
      })
      .catch(error => {
        if (error.response) {
          const { status } = error.response;
          if (status === 404) {
            this.setState({
              orders: []
            });
          }
        } else if (error.request) {
          notify.show("Wrong request", "error", 4000);
        }
      });
  };

  componentWillMount() {
    this.getMenu();
  }
  render() {
    let daysList = [
      { dataDay: "monday", value: "Monday" },
      { dataDay: "tuesday", value: "Tuesday" },
      { dataDay: "wednesday", value: "Wednesday" },
      { dataDay: "thursday", value: "Thursday" },
      { dataDay: "friday", value: "Friday" },
      { dataDay: "saturday", value: "Saturday" },
      { dataDay: "sunday", value: "Sunday" }
    ];
    let days = daysList.map((day, i) => (
      <li key={i} className="list-group-item">
        <button className="btn btn-block btn-lg" data-day={day.dataDay}>
          {day.value}
        </button>
      </li>
    ));

    const { menu, loaded } = this.state;
    const menuDetails =
      menu.length === 0 ? (
        <div>No Menu Found</div>
      ) : (
        menu.map(menumeal => (
          <div key={menumeal.id}>
            <Menu
              id={menumeal.id}
              mealName={menumeal.meal_name}
              price={menumeal.price}
              key={menumeal.id}
              handleAddOrder={this.handleAddOrder}
            />
          </div>
        ))
      );

    return (
      <div className="row">
        <div className="col-md-2">
          <div className="header">Days</div>
          <ul className="list-group">{days}</ul>
        </div>
        <div className="col-md-6 meal-list">
          <h2 className="header text-center">Menu List</h2>
          <Loader loaded={loaded}>
            <div className="row">{menuDetails}</div>
          </Loader>
        </div>
        <div className="col-md-4">
          <div className="ordered">
            <div className="header text-center">Order</div>
            {/* <Loader loaded={loaded}>
              <div className="row">{Orders}</div>
            </Loader> */}
          </div>
        </div>
      </div>
    );
  }
}
export default AdminMenu;
