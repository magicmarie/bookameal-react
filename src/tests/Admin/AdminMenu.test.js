import React from "react";
import { shallow } from "enzyme";
import MockAdapter from "axios-mock-adapter";
import AdminMenu from "../../components/Admin/AdminMenu";
import axiosInstance from "../../components/common/Apicalls";
import getCurrentDay from "../../components/common/CurrentDay";

describe("AdminMenu component", () => {
  let data;
  let wrapper;
  beforeEach(() => {
    data = [
      {
        id: 1,
        meals: [{ id: 1, meal_name: "test", price: 1000 }],
        name: getCurrentDay()
      }
    ];
    wrapper = shallow(<AdminMenu />);
  });
  it("renders successfully", async () => {
    const axiosMock = new MockAdapter(axiosInstance);
    await axiosMock.onGet("/admin-menus").reply(200, {
      menus: data
    });
  });
  it("it catches failures", () => {
    const axiosMock = new MockAdapter(axiosInstance);
    axiosMock.onGet("/admin-menus").reply(401, {
      menus: data
    });
  });
  it("manipulates array to get menu object", () => {
    const today = getCurrentDay();
    wrapper.instance().setState({ menus: data });
    wrapper.instance().getMenu(today);
    const { currentMenu } = wrapper.instance().state;
    expect(currentMenu).toEqual(data[0]);
  });

  it("should confirm delete", () => {
    const id = 1;
    const axiosMock = new MockAdapter(axiosInstance);
    axiosMock.onDelete(`/menu/${id}`).reply(200, {
      message: "success"
    });
    const modal = document.createElement("div");
    modal.setAttribute("id", `cancelModal${id}`);
    document.body.appendChild(modal);
    wrapper.instance().ConfirmDeleteMeal(id);
  });
  it("renders successfully", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
