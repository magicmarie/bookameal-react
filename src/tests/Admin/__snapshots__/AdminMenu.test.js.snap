// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminMenu component renders successfully 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AdminMenu />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Unknown
          getMenu={[Function]}
/>,
        <div
          className="vl"
/>,
        <div
          className="col-md-9 meal-list"
>
          <h2
                    className="header text-center"
          >
                    
                    's Menu List
          </h2>
          <Unknown
                    component="div"
                    loaded={false}
                    loadedClassName="loadedContent"
                    parentClassName="loader"
          >
                    <div
                              className="row"
                    >
                              <div>
                                        No Menu Found
                              </div>
                    </div>
          </Unknown>
</div>,
      ],
      "className": "row",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "getMenu": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "vl",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h2
              className="header text-center"
>
              
              's Menu List
</h2>,
            <Unknown
              component="div"
              loaded={false}
              loadedClassName="loadedContent"
              parentClassName="loader"
>
              <div
                            className="row"
              >
                            <div>
                                          No Menu Found
                            </div>
              </div>
</Unknown>,
          ],
          "className": "col-md-9 meal-list",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "",
                "'s Menu List",
              ],
              "className": "header text-center",
            },
            "ref": null,
            "rendered": Array [
              "",
              "'s Menu List",
            ],
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <div
                className="row"
>
                <div>
                                No Menu Found
                </div>
</div>,
              "component": "div",
              "loaded": false,
              "loadedClassName": "loadedContent",
              "parentClassName": "loader",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div>
                  No Menu Found
</div>,
                "className": "row",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "No Menu Found",
                },
                "ref": null,
                "rendered": "No Menu Found",
                "type": "div",
              },
              "type": "div",
            },
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Unknown
            getMenu={[Function]}
/>,
          <div
            className="vl"
/>,
          <div
            className="col-md-9 meal-list"
>
            <h2
                        className="header text-center"
            >
                        
                        's Menu List
            </h2>
            <Unknown
                        component="div"
                        loaded={false}
                        loadedClassName="loadedContent"
                        parentClassName="loader"
            >
                        <div
                                    className="row"
                        >
                                    <div>
                                                No Menu Found
                                    </div>
                        </div>
            </Unknown>
</div>,
        ],
        "className": "row",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "getMenu": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "vl",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h2
                className="header text-center"
>
                
                's Menu List
</h2>,
              <Unknown
                component="div"
                loaded={false}
                loadedClassName="loadedContent"
                parentClassName="loader"
>
                <div
                                className="row"
                >
                                <div>
                                                No Menu Found
                                </div>
                </div>
</Unknown>,
            ],
            "className": "col-md-9 meal-list",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "",
                  "'s Menu List",
                ],
                "className": "header text-center",
              },
              "ref": null,
              "rendered": Array [
                "",
                "'s Menu List",
              ],
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <div
                  className="row"
>
                  <div>
                                    No Menu Found
                  </div>
</div>,
                "component": "div",
                "loaded": false,
                "loadedClassName": "loadedContent",
                "parentClassName": "loader",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <div>
                    No Menu Found
</div>,
                  "className": "row",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "No Menu Found",
                  },
                  "ref": null,
                  "rendered": "No Menu Found",
                  "type": "div",
                },
                "type": "div",
              },
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
